File: apps/api/src/sync/upsert-engine.ts
import { SheetsClient } from './sheets-client.ts';
import { SyncDiffer } from './differ.ts';
import { WorksheetMapper, RowMappingConfig, ParsedAttendanceRow } from './worksheet-mapper.ts';
import { SyncLogger } from './sync-logger.ts';

export class UpsertEngine {
  private sheetsClient: SheetsClient;
  private syncLogger: SyncLogger;
  private prismaClient: any;
  private inMemoryDb: Map<string, ParsedAttendanceRow & { id: string; updatedAt: Date }> = new Map();
  private historyEvents: Array<{ recordId: string; prevStatus: string | null; newStatus: string; timestamp: Date }> = [];

  constructor(sheetsClient?: SheetsClient, prismaClient?: any) {
    this.sheetsClient = sheetsClient || new SheetsClient();
    this.prismaClient = prismaClient || null;
    this.syncLogger = new SyncLogger(this.prismaClient);
  }

  /**
   * Parse 2D Sheets values into structured attendance entities via WorksheetMapper
   */
  parseWorksheetRows(values: string[][], config: RowMappingConfig): { validRows: ParsedAttendanceRow[]; skippedCount: number } {
    if (!values || values.length <= 1) {
      return { validRows: [], skippedCount: 0 };
    }

    const validRows: ParsedAttendanceRow[] = [];
    let skippedCount = 0;

    // Skip header row
    for (let i = 1; i < values.length; i++) {
      const row = values[i];
      const sourceRowHash = SyncDiffer.computeRowHash(row);
      const entity = WorksheetMapper.mapRowToEntity(row, config, sourceRowHash);

      if (entity) {
        validRows.push(entity);
      } else {
        skippedCount++;
      }
    }

    return { validRows, skippedCount };
  }

  /**
   * Idempotent batch sync execution
   */
  async executeSyncRun(
    sheetId: string,
    range: string,
    config: RowMappingConfig,
    lastContentHash?: string,
    worksheetMappingId?: string
  ) {
    const logEntry = this.syncLogger.startRun(worksheetMappingId);

    try {
      const rawValues = await this.sheetsClient.getRangeValues(sheetId, range);
      const newContentHash = SyncDiffer.computeRangeHash(rawValues);

      // Range-level hash short-circuit (PRD §6.2)
      if (lastContentHash && newContentHash === lastContentHash) {
        await this.syncLogger.finishRun(logEntry, {
          status: 'SKIPPED_NO_CHANGE',
          rowsRead: rawValues.length,
          rowsUpserted: 0,
          contentHash: newContentHash
        });
        return {
          status: 'SKIPPED_NO_CHANGE',
          rowsRead: rawValues.length,
          rowsUpserted: 0,
          contentHash: newContentHash,
          historyEventsCreated: 0
        };
      }

      const { validRows, skippedCount } = this.parseWorksheetRows(rawValues, config);
      let updatedCount = 0;
      let historyEventsCount = 0;

      // Upsert valid rows transactional batch
      for (const row of validRows) {
        const naturalKey = `${row.studentRollNo}_${row.lectureDate.toISOString().split('T')[0]}`;
        const existingRecord = this.inMemoryDb.get(naturalKey);

        if (this.prismaClient?.attendanceRecord) {
          try {
            // Prisma natural key transactional upsert
            const dbRecord = await this.prismaClient.attendanceRecord.findFirst({
              where: {
                student: { rollNo: row.studentRollNo },
                lectureDate: row.lectureDate
              }
            });

            if (!dbRecord) {
              const created = await this.prismaClient.attendanceRecord.create({
                data: {
                  student: { connect: { rollNo: row.studentRollNo } },
                  subject: { connect: { id: worksheetMappingId || 'subj_default' } },
                  lectureDate: row.lectureDate,
                  status: row.status,
                  faculty: row.faculty,
                  remarks: row.remarks,
                  sourceRowHash: row.sourceRowHash
                }
              });
              await this.prismaClient.attendanceHistoryEvent.create({
                data: {
                  attendanceRecordId: created.id,
                  previousStatus: null,
                  newStatus: row.status,
                  syncLogId: logEntry.id
                }
              });
              updatedCount++;
              historyEventsCount++;
            } else if (dbRecord.sourceRowHash !== row.sourceRowHash || dbRecord.status !== row.status) {
              await this.prismaClient.attendanceRecord.update({
                where: { id: dbRecord.id },
                data: {
                  status: row.status,
                  faculty: row.faculty,
                  remarks: row.remarks,
                  sourceRowHash: row.sourceRowHash
                }
              });
              await this.prismaClient.attendanceHistoryEvent.create({
                data: {
                  attendanceRecordId: dbRecord.id,
                  previousStatus: dbRecord.status,
                  newStatus: row.status,
                  syncLogId: logEntry.id
                }
              });
              updatedCount++;
              historyEventsCount++;
            }
          } catch (dbErr) {
            console.warn('UpsertEngine: DB transactional upsert fallback to in-memory store:', dbErr);
          }
        } else {
          // In-memory fallback upsert logic
          if (!existingRecord) {
            const recordId = `att_${Date.now()}_${Math.random().toString(36).substr(2, 4)}`;
            this.inMemoryDb.set(naturalKey, { id: recordId, updatedAt: new Date(), ...row });
            this.historyEvents.push({ recordId, prevStatus: null, newStatus: row.status, timestamp: new Date() });
            updatedCount++;
            historyEventsCount++;
          } else if (existingRecord.sourceRowHash !== row.sourceRowHash || existingRecord.status !== row.status) {
            const prevStatus = existingRecord.status;
            this.inMemoryDb.set(naturalKey, { ...existingRecord, updatedAt: new Date(), ...row });
            this.historyEvents.push({ recordId: existingRecord.id, prevStatus, newStatus: row.status, timestamp: new Date() });
            updatedCount++;
            historyEventsCount++;
          }
        }
      }

      const finalStatus = skippedCount > 0 ? 'PARTIAL_FAILURE' : 'SUCCESS';
      const errorMessage = skippedCount > 0 ? SyncLogger.formatErrorReason('MALFORMED_ROWS', `${skippedCount} row(s) skipped due to missing date/roll number/status`) : undefined;

      await this.syncLogger.finishRun(logEntry, {
        status: finalStatus,
        rowsRead: rawValues.length,
        rowsUpserted: updatedCount,
        contentHash: newContentHash,
        errorMessage
      });

      return {
        status: finalStatus,
        rowsRead: rawValues.length,
        rowsUpserted: updatedCount,
        skippedRows: skippedCount,
        contentHash: newContentHash,
        processedRecords: validRows,
        historyEventsCreated: historyEventsCount
      };
    } catch (error: any) {
      const is403 = error?.message?.includes('403');
      const is429 = error?.message?.includes('429');
      const errType = is403 ? '403_FORBIDDEN' : is429 ? '429_RATE_LIMITED' : 'DB_ERROR';
      const errorMessage = SyncLogger.formatErrorReason(errType, error?.message);

      await this.syncLogger.finishRun(logEntry, {
        status: 'FAILED',
        rowsRead: 0,
        rowsUpserted: 0,
        errorMessage
      });

      throw error;
    }
  }

  getInMemoryRecords() {
    return Array.from(this.inMemoryDb.values());
  }

  getHistoryEvents() {
    return this.historyEvents;
  }
}

File: backend/src/main/java/com/vesit/openattend/service/sync/UpsertEngine.java
package com.vesit.openattend.service.sync;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.vesit.openattend.entity.*;
import com.vesit.openattend.entity.enums.AttendanceStatus;
import com.vesit.openattend.entity.enums.SyncRunStatus;
import com.vesit.openattend.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.IOException;
import java.util.*;

@Slf4j
@Service
@RequiredArgsConstructor
public class UpsertEngine {

    private final SheetsClient sheetsClient;
    private final SyncLogger syncLogger;
    private final StudentRepository studentRepository;
    private final AttendanceRecordRepository attendanceRecordRepository;
    private final AttendanceHistoryEventRepository attendanceHistoryEventRepository;
    private final WorksheetMappingRepository worksheetMappingRepository;
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Transactional
    public SyncResult executeSyncRun(WorksheetMapping mapping, List<List<Object>> overrideValues, String lastContentHash) {
        SyncLog syncLog = syncLogger.startRun(mapping);

        try {
            List<List<Object>> rawValues = overrideValues;
            if (rawValues == null) {
                rawValues = sheetsClient.getRangeValues(mapping.getSheetId(), mapping.getRange());
            }

            if (rawValues == null || rawValues.isEmpty()) {
                syncLogger.finishRun(syncLog, SyncRunStatus.SUCCESS, 0, 0, null, null);
                return SyncResult.builder()
                        .status(SyncRunStatus.SUCCESS)
                        .rowsRead(0)
                        .rowsUpserted(0)
                        .build();
            }

            String newContentHash = SyncDiffer.computeRangeHash(rawValues);

            // Range-level hash short-circuit (PRD §6.2)
            if (lastContentHash != null && lastContentHash.equals(newContentHash)) {
                syncLogger.finishRun(syncLog, SyncRunStatus.SKIPPED_NO_CHANGE, rawValues.size(), 0, newContentHash, null);
                return SyncResult.builder()
                        .status(SyncRunStatus.SKIPPED_NO_CHANGE)
                        .rowsRead(rawValues.size())
                        .rowsUpserted(0)
                        .contentHash(newContentHash)
                        .historyEventsCreated(0)
                        .build();
            }

            // Parse Column Roles from Mapping JSON
            Map<String, String> columnRoles = Collections.emptyMap();
            try {
                if (mapping.getColumnRoles() != null && !mapping.getColumnRoles().trim().isEmpty()) {
                    columnRoles = objectMapper.readValue(mapping.getColumnRoles(), new TypeReference<Map<String, String>>() {});
                }
            } catch (Exception e) {
                log.warn("Failed to parse columnRoles JSON for mapping {}: {}", mapping.getId(), e.getMessage());
            }
            RowMappingConfig config = RowMappingConfig.builder().columnRoles(columnRoles).build();

            int updatedCount = 0;
            int historyEventsCount = 0;
            int skippedCount = 0;
            List<String> skipReasons = new ArrayList<>();

            Subject subject = mapping.getSubject();

            // Skip header row (index 0)
            for (int i = 1; i < rawValues.size(); i++) {
                List<Object> row = rawValues.get(i);
                String sourceRowHash = SyncDiffer.computeRowHash(row);
                ParsedAttendanceRow parsedRow = WorksheetMapper.mapRowToEntity(row, config, sourceRowHash);

                if (parsedRow == null) {
                    skippedCount++;
                    skipReasons.add("Row " + (i + 1) + ": Malformed date or missing roll number");
                    continue;
                }

                Optional<Student> studentOpt = studentRepository.findByRollNo(parsedRow.getStudentRollNo());
                if (studentOpt.isEmpty()) {
                    skippedCount++;
                    skipReasons.add("Row " + (i + 1) + ": Unknown student roll number [" + parsedRow.getStudentRollNo() + "]");
                    continue;
                }

                Student student = studentOpt.get();

                Optional<AttendanceRecord> existingOpt = attendanceRecordRepository
                        .findByStudentIdAndSubjectIdAndLectureDateAndSessionIndex(
                                student.getId(),
                                subject.getId(),
                                parsedRow.getLectureDate(),
                                parsedRow.getSessionIndex()
                        );

                if (existingOpt.isEmpty()) {
                    // Create new attendance record
                    AttendanceRecord newRecord = AttendanceRecord.builder()
                            .id(UUID.randomUUID().toString())
                            .student(student)
                            .subject(subject)
                            .lectureDate(parsedRow.getLectureDate())
                            .sessionIndex(parsedRow.getSessionIndex())
                            .status(parsedRow.getStatus())
                            .faculty(parsedRow.getFaculty())
                            .remarks(parsedRow.getRemarks())
                            .sourceRowHash(parsedRow.getSourceRowHash())
                            .build();
                    attendanceRecordRepository.save(newRecord);

                    // Create initial history event
                    AttendanceHistoryEvent event = AttendanceHistoryEvent.builder()
                            .id(UUID.randomUUID().toString())
                            .attendanceRecord(newRecord)
                            .previousStatus(null)
                            .newStatus(parsedRow.getStatus())
                            .syncLogId(syncLog.getId())
                            .build();
                    attendanceHistoryEventRepository.save(event);

                    updatedCount++;
                    historyEventsCount++;
                } else {
                    AttendanceRecord existing = existingOpt.get();
                    // Row-level short-circuit: update only if hash or status changed
                    if (!Objects.equals(existing.getSourceRowHash(), parsedRow.getSourceRowHash()) ||
                            existing.getStatus() != parsedRow.getStatus()) {

                        AttendanceStatus prevStatus = existing.getStatus();
                        existing.setStatus(parsedRow.getStatus());
                        existing.setFaculty(parsedRow.getFaculty());
                        existing.setRemarks(parsedRow.getRemarks());
                        existing.setSourceRowHash(parsedRow.getSourceRowHash());
                        attendanceRecordRepository.save(existing);

                        AttendanceHistoryEvent event = AttendanceHistoryEvent.builder()
                                .id(UUID.randomUUID().toString())
                                .attendanceRecord(existing)
                                .previousStatus(prevStatus)
                                .newStatus(parsedRow.getStatus())
                                .syncLogId(syncLog.getId())
                                .build();
                        attendanceHistoryEventRepository.save(event);

                        updatedCount++;
                        historyEventsCount++;
                    }
                }
            }

            SyncRunStatus finalStatus = skippedCount > 0 ? SyncRunStatus.PARTIAL_FAILURE : SyncRunStatus.SUCCESS;
            String errorMessage = skippedCount > 0
                    ? SyncLogger.formatErrorReason("MALFORMED_ROWS", String.join("; ", skipReasons))
                    : null;

            syncLogger.finishRun(syncLog, finalStatus, rawValues.size(), updatedCount, newContentHash, errorMessage);

            return SyncResult.builder()
                    .status(finalStatus)
                    .rowsRead(rawValues.size())
                    .rowsUpserted(updatedCount)
                    .skippedRows(skippedCount)
                    .contentHash(newContentHash)
                    .historyEventsCreated(historyEventsCount)
                    .errorMessage(errorMessage)
                    .build();

        } catch (Exception e) {
            log.error("Sync run failed with error: {}", e.getMessage(), e);
            String errReason = SyncLogger.formatErrorReason("SYNC_ERROR", e.getMessage());
            syncLogger.finishRun(syncLog, SyncRunStatus.FAILED, 0, 0, null, errReason);
            throw new RuntimeException("Sync execution failed: " + e.getMessage(), e);
        }
    }
}
